feat(eval): implement Docker-based execution environment

TimeToBuildBob opened the pull request:
## Overview

Implements Docker-based execution environment for GEPA evaluations to prevent host environment pollution (e.g., git config changes).

## Changes

- **Add DockerExecutionEnv class** to `gptme/eval/execenv.py`
  - Container lifecycle management (start, cleanup, automatic)
  - Volume-based file transfer for efficiency
  - Same interface as SimpleExecutionEnv (backward compatible)
  - 30-second timeout (matches existing behavior)

- **Add --use-docker CLI flag** to `gptme-eval`
  - Enables Docker-based execution for isolation
  - Default: SimpleExecutionEnv (backward compatible)

- **Integration** into eval system
  - Parameter flow: main → run_evals → execute
  - Conditional execution based on flag

## Usage

```bash
# Build Docker image (one-time)
make build-docker

# Run with Docker isolation
gptme-eval hello --model anthropic/claude-haiku-4-5 --use-docker

# Run without Docker (default, backward compatible)
gptme-eval hello --model anthropic/claude-haiku-4-5
```

## Benefits

- ✅ Environment isolation (prevents host pollution)
- ✅ Clean execution per test
- ✅ Reproducible results
- ✅ Backward compatible

## Testing

- ✅ Syntax check passed
- ✅ Type check (mypy) passed
- ✅ Pre-commit hooks passed
- ⏭️ Runtime testing needed

Addresses ErikBjare/bob#73
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Introduces `DockerExecutionEnv` for isolated evaluations using Docker, with a `--use-docker` flag for optional execution environment isolation.
> 
>   - **Behavior**:
>     - Adds `DockerExecutionEnv` class in `execenv.py` for isolated command execution using Docker.
>     - Integrates Docker execution into the evaluation system with a `--use-docker` flag in `main.py`.
>     - Maintains backward compatibility with `SimpleExecutionEnv`.
>   - **Execution Environment**:
>     - `DockerExecutionEnv` manages container lifecycle, file transfers, and command execution.
>     - Implements a 30-second timeout for command execution.
>   - **CLI Changes**:
>     - Adds `--use-docker` flag to `gptme-eval` in `main.py` to enable Docker-based execution.
>   - **Integration**:
>     - Modifies `run_evals` and `execute` functions in `run.py` to conditionally use Docker based on the flag.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=gptme%2Fgptme&utm_source=github&utm_medium=referral)<sup> for 17de4915a2ed731919c32dc3d1c40b6d6340b443. You can [customize](https://app.ellipsis.dev/gptme/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

ErikBjare commented:
Right, but how does this actually help the GEPA? This doesn't seem to affect the GEPA evaluations?

TimeToBuildBob commented:
## Addressed Remaining Greptile Concerns

Fixed 3 of 4 issues flagged by Greptile (resource leak already fixed in 6e91b586):

### ✅ Fix 1: Binary Encoding Consistency
**Issue**: `download()` returned base64 string instead of bytes, breaking Files type contract
**Fix**: Removed `.decode()` to return base64-encoded bytes matching FileStore contract
- Before: `base64.b64encode(f.read()).decode()`
- After: `base64.b64encode(f.read())`

### ✅ Fix 2: Error Handling for Missing Docker Image
**Issue**: No helpful guidance when Docker image doesn't exist
**Fix**: Added try-except around `docker run` with specific error messages
- Catches `CalledProcessError`
- Detects "Unable to find image" / "No such image"
- Suggests: `make build-docker`

### ✅ Fix 3: Complete Timeout Handling
**Issue**: Timeout killed local process but container command continued running
**Fix**: Added `docker stop` on timeout to terminate container
- Stops container with 5-second timeout
- Ensures no orphaned processes

### ✅ Fix 4: Resource Leak (Previously Fixed)
**Fixed in 6e91b586**: Added try-finally cleanup in `execute()` function

All fixes verified:
- ✅ Python compilation passed
- ✅ Pre-commit hooks passed (ruff, mypy)
- ✅ Code matches FileStore implementation patterns

**Commit**: 17de4915
